plugins/woocommerce/tests/php/includes/gateways/paypal/class-wc-gateway-paypal-test.php

```diff
@@ -223,4 +223,69 @@ public function test_correct_settings_is_displayed_when_orders_v2_is_disabled()
 			}
 		}
 	}
+
+	/**
+	 * Test that gateway is available when Orders v2 is disabled (legacy mode).
+	 */
+	public function test_is_available_with_legacy_mode() {
+		// Enable the gateway.
+		update_option( 'woocommerce_paypal_settings', array( 'enabled' => 'yes' ) );
+
+		// Mock Orders v2 to be disabled.
+		$mock_gateway = $this->getMockBuilder( WC_Gateway_Paypal::class )
+			->onlyMethods( array( 'should_use_orders_v2' ) )
+			->getMock();
+		$mock_gateway->method( 'should_use_orders_v2' )->willReturn( false );
+
+		$this->assertTrue( $mock_gateway->is_available() );
+	}
+
+	/**
+	 * Test that gateway is not available when Orders v2 is enabled and email is empty.
+	 */
+	public function test_is_available_with_orders_v2_and_no_email() {
+		// Enable the gateway with empty email.
+		update_option(
+			'woocommerce_paypal_settings',
+			array(
+				'enabled' => 'yes',
+				'email'   => '',
+			)
+		);
+
+		// Mock Orders v2 to be enabled.
+		$mock_gateway = $this->getMockBuilder( WC_Gateway_Paypal::class )
+			->onlyMethods( array( 'should_use_orders_v2' ) )
+			->getMock();
+		$mock_gateway->method( 'should_use_orders_v2' )->willReturn( true );
+
+		// Gateway should not be available with empty email in Orders v2.
+		$this->assertFalse( $mock_gateway->is_available() );
+	}
```

Comment:
While I love dataproviders, it feels like that might be a good idea to use something like that to cover the following cases:
* The `email` field is not set at all -> should not be available
* The `email` field is `''` -> should not be available
* The `email` field has an invalid email address, e.g. `example@invalid`
* The `email` field is a valid email address

We could rewrite this test something like the following to make additional test cases easy to add.
```suggestion
	public function test_is_available_with_orders_v2( ?string $email, bool $expected_value ) {
		$new_settings = array(
			'enabled' => 'yes',
		);
		if ( null !== $email ) {
			 $new_settings['email'] = $email;
		}
		update_option( 'woocommerce_paypal_settings', $new_settings );

		// Mock Orders v2 to be enabled.
		$mock_gateway = $this->getMockBuilder( WC_Gateway_Paypal::class )
			->onlyMethods( array( 'should_use_orders_v2' ) )
			->getMock();
		$mock_gateway->method( 'should_use_orders_v2' )->willReturn( true );

		$this->assertSame( $expected_value, $mock_gateway->is_available() );
	}
```

Reply:
Addressed in [ae38ae7](https://github.com/woocommerce/woocommerce/pull/62017/commits/ae38ae7f42c6eb9a2864d77605735527051510d4).

Skipped the case when `The email field is not set at all` as the email field will always be present as it is coming from the settings which defaults to the `admin_email`.

Reply:
I think it's worth adding a defensive test case. While that _shouldn't_ happen it can happen, so why note make sure we're testing our implementation for that edge case. `wp option patch delete` can mess things up for anyone! 😉

Reply:
Added in [41558f2](https://github.com/woocommerce/woocommerce/pull/62017/commits/41558f22aac0d54790bc76b14fe3de74f25aa914)